kernel/src/sched/sched_class/mod.rs

```diff
@@ -0,0 +1,227 @@
+
+        let cpu_affinity = thread.lock_cpu_affinity();
+        let cpu = self.select_cpu(&cpu_affinity);
+        task.schedule_info().cpu.set_if_is_none(cpu).ok()?;
+        drop(cpu_affinity);
+
+        let mut rq = self.rqs[cpu as usize].disable_irq().lock();
+        rq.enqueue_thread(thread);
+        Some(cpu)
+    }
+
+    fn local_mut_rq_with(&self, f: &mut dyn FnMut(&mut dyn LocalRunQueue)) {
+        let guard = disable_local();
+        let mut lock = self.rqs[guard.current_cpu() as usize].lock();
+        f(&mut *lock)
+    }
+
+    fn local_rq_with(&self, f: &mut dyn FnMut(&dyn LocalRunQueue)) {
+        let guard = disable_local();
+        f(&*self.rqs[guard.current_cpu() as usize].lock())
+    }
+}
+
+impl ClassScheduler {
+    fn select_cpu(&self, affinity: &CpuSet) -> u32 {
+        let guard = disable_local();
+        let cur = guard.current_cpu();
+        if affinity.contains(cur) {
+            cur
+        } else {
+            affinity.iter().next().expect("empty affinity")
+        }
+    }
+}
+
+impl PerCpuClassRqSet {
+    fn pick_next_thread(&mut self) -> Option<Arc<Thread>> {
+        (self.stop.pick_next())
+            .or_else(|| self.real_time.pick_next())
+            .or_else(|| self.fair.pick_next())
+            .or_else(|| self.idle.pick_next())
+    }
+
+    fn enqueue_thread(&mut self, thread: &Arc<Thread>) {
+        let cloned = thread.clone();
+        let entity = thread.sched_entity().lock();
+        match *entity {
+            SchedEntity::Stop(_) => self.stop.enqueue(cloned, entity),
+            SchedEntity::RealTime(_) => self.real_time.enqueue(cloned, entity),
+            SchedEntity::Fair(_) => self.fair.enqueue(cloned, entity),
+            SchedEntity::Idle(_) => self.idle.enqueue(cloned, entity),
+        };
+    }
+}
+
+impl LocalRunQueue for PerCpuClassRqSet {
+    fn current(&self) -> Option<&Arc<Task>> {
+        self.current.as_ref()
+    }
+
+    fn pick_next_current(&mut self) -> Option<&Arc<Task>> {
+        match self.pick_next_thread() {
+            Some(next) => {
+                let next_task = next.task();
+                if let Some(old_task) = self.current.replace(next_task.upgrade().unwrap()) {
+                    if Arc::as_ptr(&old_task) == Weak::as_ptr(next_task) {
+                        return None;
+                    }
+                    let old = Thread::borrow_from_task(&old_task);
+                    self.enqueue_thread(old);
+                }
+                self.current.as_ref()
+            }
+            None => None,
+        }
+    }
+
+    fn update_current(&mut self, flags: UpdateFlags) -> bool {
+        if let Some(cur_task) = &self.current {
+            let cur = Thread::borrow_from_task(cur_task);
+            match &mut *cur.sched_entity().lock() {
+                SchedEntity::Stop(stop_entity) => self.stop.update_current(stop_entity, flags),
+                SchedEntity::RealTime(real_time_entity) => {
```

Comment:
I don't think the current implementation of `update_current` is correct since it does not respect the scheduling priority. Let's say the current task belongs to the fair scheduler. What if a timer interrupt occurs and a real-time thread is runnable? The real-time runqueue is not even looked.